### ISSUE TITLE
Crash with recursive conditional type during serialization

### ISSUE THREAD

**Issue by jBernavaPrah** [microsoft/TypeScript#43529]
# Bug Report
Hi guys,
I am not sure if it is a bug but it is a strange behaviour surely. 
I have created some Types that help me during the development of my ReactJs app. 

I can use those types separately and they show output result quickly and use the normal quantity of memory.

Although, when I try to reuse those types with other Types, the compilation of typescript on my ReactJs app is not completed and after some time the node that runs `yarn start` crashes because it has reached the maximal memory allowed (it can be 2Gb or 8Gb, it is just a matter of time before the crash). If I comment out those types, the ReactJs compilation is completed as expected. 

I think I have recreated the same behaviour also on Playground. In this case, the typescript worker used by Chrome will grow to 1.5Gb, without showing information on the popup about the Type.

I have created this [Video](https://youtu.be/6Az7PX6kSdg) to understand better what I mean by it.

On the playground I have put what I think is the relevant code from my project, to show this error.

### 🔎 Search Terms

typescript, recursion, JavaScript heap out of memory, ReactJs

### 🕗 Version & Regression Information

Playground and typescript npm 4.2.5 and Nightly.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210401#code/KYDwDg9gTgLgBAbwL4CgD0AqDK7d3AZQgFtg4ALYAGzGCjgBUBPWgZzgEMA7AEzgEkuMOgDMOAY2CsAdNJx4MaFClCRYcGCzIBhbgCFgAUXDcewHgB4GcUMN7sIAIwBWwcfAC8cJ6-cAaOAAROC9WGCgASy4AcwA+EMQcOGSANw4qAFdgAC5GJOSzMQyqGFzAlFQUKOEoMUk4PQ5WMgRlSurRCTIAVWb6W2B7BqaW-LgInlywyJi-MagIKhy4XS4DYzBTcwsAJUXgWLnk5K4IYSnwqOi59qFO+r2lmxA7HnZG5sTj8cm4aaujscMn1cqt1iZeNtenRDmMQBcZtEKso0GhGOQIuwOrUunAAO4RKhUOCOMiqLZ8KJwcQRKDiYoceh4jhMOTYupkAAaz1e7xGX2OEwRALGXA4pGFszGwLooP0RghZks0KgsO+CyWcrWCs2kMsjwO832rC14N1St2+wA2gBdWLIlCaWhwABSECiFgA0gEAArxLyenmDN5-S4xOAAHzgXAyxFJ9AA-GMfUGhv9w1GY3G6HAk99vgADAAkCE9SBLACIK6mQymE3Aq3BchXpBXywgfUgC2NvrkuMAUjm+wO6ABuZROsg+qDAEQREAAdQiMHIFh9AQIAW0CQA5NId-64CmBkMdzvc3Biwh09E4AAyQhdpuXks3+9H9tvh-aT9h28PgguwnLRCDAKhlyXFcAEEiR9CBWFYCJHCWbQSEcKIOBgCIIC4VgLAIGt2BvAJghPEMb39MY3zI9gCPrN9sh7Y4CJouAzwvdjGPzbjCEIl8ECiEQcwYdtAnbQSc26J88x42TrCjN0PQYDcwIg5dyBgqg4IQpCULQjCsJwvDuhI2J7Vk2TcgIccUSweR0UxOARAyLh3Gwrh8UJYkZxgDIoA89JiUgHT0KeABrYAmHYFdMOpbgSTJRVzG8DzlJSqgmA0CANEoOAADk4DMYAwDgJZByoeQlEnOBPUi1gwR1CkAH0rACArWKzeMSOKlc+M6uhbUPdqXmDdhAh68grR3JYYhXHcbQvftB3oLjjmsViGo2ClcyY5Jaqizakp4FqGCmtJMmAea2u6sAV3iVbuPWkahigqAoBZCwJPobp4nrfb6vlLa9Rakz8puu6m2UCy1r4nw3HgGToe+VokYsq1AypCKmAgERGDgBaHtR-NTs9BbWLh9wdqJpHA1Yt9M1jeNdupnj6yvMsC0jV13S4Cx2a7AJ-sO81zBO9GbWuuArQARgCWRpHG27yDtcyWdR4dluZtWNboKGWaQK0sZxxgbS1njCdRpbdZUcBoHgaqhcBo7WvyvrGZzLwABZD0d7UgaVE7JcG2yFDgf24CxhwPILAB5TnHBGPgcMvf7Oc2d7SBqORcCqkD-aalqY8FurfT44jXY6936HB8g+JcsLTjxLhbTgSjvmG3kggmqaZuiOaFsR-MY744XtsH2SY6mwoOGKGB5sh1Hh4t45h42p2Rb4ceePzwurWPZ6Q3+jjzqyc9ch3afZ6umqS6PSXFYh5fuNXg-2Fe96mE+rghO+36ze4neFhQb-V9PfCasRbT-3zLkF+ncKYIygcjRBPF0bjA8kbXGMcbQAFoEy5EAZPUmxcoq+hnHORc6k1yC0OGDKWss4DywfsrVW1NUDazgDHZQwFnQEFUjAKC2gvSHl4eBGAkENKwXgohZCwBUJxgMu5PCtNX6hkRBeQM58dwBD3AePiX0arZT+tlTRjo86KkDjVNqbtsz0C9oeQBaURHLgEUI7RWjaHB2qppf6LsO6jWjFXBI3sEi+zNBSXx8RWL6J2BeGJOsoBcPQHZfA-Bcb8G8LQDyK4yCQDABkMAAQAAUkBqhXE8r1bJcBiAQBlClHKZB-YAEoAiVPECQMAhIcysHIBAPE0VcrVX9gEckepODsFUPDcw2dQ78DnuwMQYROC8HGPAHgEApBcB3PAWpHAeA8GXO5dIVTgDVKgEweQlUbZqHtmY80VhYYuHhoLPi3i6oRISFbKAVjK42KCfY8xLtvT5XtEk0ODBKCfBgHibKzlXKGVwviaAYUxnUnIIyYgAR0isGytkzKjIyCsHyZ0RZpgUUTPcMlXFhUcJbLgDsvZBycJHIALInOgOcnOKAljwCYNLfBioLCcntNyuATAABM-K7mcm0TKKA0gNTAGkLK+V+wDxcMwAoey8towjlgJQJYCF8SUA8uk-slLsq1MqeHKkZxKD0GYGwTF-SyAiEWFQXpZTQDijAlIFpuVcn5L+D0vEDZ3W7KuPLaspKCQrhqfAbpvSlmZUEtAYgmF3JLJ4PZSprLTmZQABLAA4CVaICw+lZTgNLaQYroiODgIUqkxBMriBGI0mwy1MpYVINqgAmjUuKHlmhkGXBkwYZTKnaHIAsUgXcUgMAgIsdgpLxCUHEMiyppA82MA4I4aZ9ksAHuSRgAQOUojroxFidgLbgRlIrAAMSggwKCAAZMOOwdgxx2LkQQs4hJuUHFUxkYVsFtOIJsdw7B+yMgoEWkq4FiAjs0hAFt8KnIcE6XwbBroOBpAIOISIt0YPFu8BkeAxtN3sorPZZOTa4BgAWL4VZflx25QLEwRkg6YCMhgAWaZuBD1HsuaY500IACKWQzkx1uoo+5rEPjAGeaxV5UUXbBP+XctKnoQXVTExJpgUn4WsDFfc55rcEj50BfaHTzRxN0H09JoyABmEzNVkjqdMBYrTejv45mHvWJeOrNbCYaYqHYUhZ6tS814CzmnIkqP0X5jhz5PnBbgLpuzBnFGexc4GMz0XQvhZKJFkFoL7KOFIyi7GGQB1-GAGQO1Cn6nPF1IhZOxtBA1A5HAbky4dxESwkSRFUAwoRouSk3GHBjlbscmELyFAIjRHIAEEdabMpmr4NimdzQ2lDBgNlBNIaA1gD3ZykVpQ0s2b05loygrtEKqVX0FVSw1VAA)

### 💻 Code

```ts
// relevant code on Playground
// from line line #129 
type UseQueryOptions<T extends Base, K extends AllKeys<T, 4> > = Expand<T, K> 

type UseQueryOptions2<T , K  > = Expand_<T, K>  // line #129
type UseQueryOptions3<T , K  > = Expand_<T, K> extends infer O ? O : never  

type ExpandResult<T,K> = Expand_<T, K> extends infer O ? O : never
type UseQueryOptions4<T , K  > = ExpandResult<T,K>
```

### 🙁 Actual behavior

Every time I try to put together multiple Types, the yarn start (`react-scripts start`) is stuck on: `Files successfully emitted, waiting for typecheck results...` and after the memory node reaches the maximal allowed (can be 2Gb or 8Gb) it crashes with: 

```
<--- Last few GCs --->

[16365:0x1046ca000]   448881 ms: Mark-sweep 2028.9 (2059.0) -> 2020.9 (2059.3) MB, 766.2 / 0.0 ms  (average mu = 0.132, current mu = 0.008) allocation failure scavenge might not succeed
[16365:0x1046ca000]   449641 ms: Mark-sweep 2028.9 (2059.3) -> 2020.9 (2059.3) MB, 754.6 / 0.0 ms  (average mu = 0.073, current mu = 0.007) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1012e4da5 node::Abort() (.cold.1) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 2: 0x1000a6239 node::Abort() [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 3: 0x1000a639f node::OnFatalError(char const*, char const*) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 4: 0x1001e9007 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 5: 0x1001e8fa3 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 6: 0x100397e95 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 7: 0x10039995a v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 8: 0x100395029 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
 9: 0x1003928c1 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
10: 0x1003a115a v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
11: 0x1003a11e1 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
12: 0x10036eb87 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
13: 0x1006ed8d8 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]
14: 0x100a7a239 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/jure.prah/.nvm/versions/node/v14.16.0/bin/node]

```

### 🙂 Expected behavior

My intention is to be able to use the recursions Type together with other types.

**Comment by andrewbranch** [comment]
I get a max call stack size error when hovering `UseQueryOptions`, but can’t trigger the out of memory crash. Could you share a project / example repo that reliably crashes this way?

**Comment by jBernavaPrah** [comment]
Thank you for your response. And sorry for the delay. 

[Here](https://github.com/jBernavaPrah/test-ts-memory) you will found the example project.

On src/query/index.ts, you will find the interfaces, types and functions that I use during development. 
On src/components/TestFetchPage.tsx, you will find an example of how I use it on my project.

The master branch contains two commits:

- The current Master [commit](https://github.com/jBernavaPrah/test-ts-memory/tree/2e696dc3f0c019549493df013dcb053931868704), that will trigger the memory crash;
- And the [commit](https://github.com/jBernavaPrah/test-ts-memory/tree/07c3649a53b9add70287c8c206a67d97d2fcbb02) that shows that, if I comment lines src/queries/index.ts:73 and src/queries/index.ts:74 and uncomment the line src/queries/index.ts:76, the crash is gone and yarn runs normally.

I hope I'm helping you with this. Otherwise, do not hesitate to ask for any additional information.

**Comment by andrewbranch** [comment]
Reduced case:

```ts
export type CanBeExpanded<T extends object> = {
  value: T
}

type Expand__<O, N, Depth> =
  N extends Depth ?
      unknown :
      O extends CanBeExpanded<any> ?
          Expand__<O['value'], N, Depth> :
          O

export type UseQueryOptions<T> = Expand__<T, 4, 2>
```

**Comment by jBernavaPrah** [comment]
Hi @andrewbranch, could you please give me a temporary solution that I could use until the v4.3.1 is fixed?

**Comment by andrewbranch** [comment]
Try using types understandable by mere mortal humans? 😅 I don’t know for sure yet because I haven’t fixed it yet. Even once I understand exactly what’s going wrong, it may be difficult to apply that knowledge back to the massive types you have.

**Comment by andrewbranch** [comment]
@jBernavaPrah, I’ve had a chance to debug this and see what’s going wrong, which led me to a workaround for you. The problem comes when we’re trying to serialize these recursive types. One way that we can avoid serializing them is if we can substitute a type alias for them. So, the more you can break these large recursive types down into smaller pieces, the more we can substitute in the names of those individual pieces and break the loop. [Here I’ve extracted one part of `Expand__` into `Expand__2`](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210401#code/KYDwDg9gTgLgBAbwL4CgD0AqDK7d3AZQgFtg4ALYAGzGCjgBUBPWgZzgEMA7AEzgEkuMOgDMOAY2CsAdNJx4MaFClCRYcGCzIBhbgCFgAUXDcewHgB4GcUMN7sIAIwBWwcfAC8cJ6-cAaOAAROC9WGCgASy4AcwA+EMQcOGSANw4qAFdgAC5GJOSzMQyqGFzAlFQUKOEoMUk4PQ5WMgRlSurRCTIAVWb6W2B7BqaW-LgInlywyJi-MagIKhy4XS4DYzBTcwsAJUXgWLnk5K4IYSnwqOi59qFO+r2lmxA7HnZG5sTj8cm4aaujscMn1cqt1iZeNtenRDmMQBcZtEKso0GhGOQIuwOrUunAAO4RKhUOCOMiqLZ8KJwcQRKDiYoceh4jhMOTYupkAAaz1e7xGX2OEwRALGXA4pGFszGwLooP0RghZks0KgsO+CyWcrWCs2kMsjwO832rC14N1St2+wA2gBdWLIlCaWhwABSECiFgA0gEAArxLyenmDN5-S4xOAAHzgXAyxFJ9AA-GMfUGhv9w1GY3G6HAk99vgADAAkCE9SBLACIK6mQymE3Aq3BchXpBXywgfUgC2NvrkuMAUjm+wO6ABuZROsg+qDAEQREAAdQiMHIFh9AQIAW0CQA5NId-64CmBkMdzvc3Biwh09E4AAyQhdpuXks3+9H9tvh-aT9h28PgguwnLRCDAKhlyXFcAEEiR9CBWFYCJHCWbQSEcKIOBgCIIC4VgLAIGt2BvAJghPEMb39MY3zI9gCPrN9sh7Y4CJouAzwvdjGPzbjCEIl8ECiEQcwYdtAnbQSc26J88x42TrCjN0PQYDcwIg5dyBgqg4IQpCULQjCsJwvDuhI2J7Vk2TcgIccUSweR0UxOARAyLh3Gwrh8UJYkZxgDIoA89JiUgHT0KeABrYAmHYFdMOpbgSTJRVzG8DzlJSqgmA0CANEoOAADk4DMYAwDgJZByoeQlEnOBPUi1gwR1CkAH0rACArWKzeMSOKlc+M6uhbUPdqXmDdhAh68grR3JYYhXHcbQvftB3oLjjmsViGo2ClcyY5Jaqizakp4FqGCmtJMmAea2u6sAV3iVbuPWkahigqAoBZCwJPobp4nrfb6vlLa9Rakz8puu6m2UCy1r4nw3HgGToe+VokYsq1AypCKmAgERGDgBaHtR-NTs9BbWLh9wdqJpHA1Yt9M1jeNdupnj6yvMsC0jV13S4Cx2a7AJ-sO81zBO9GbWuuArQARgCWRpHG27yDtcyWdR4dluZtWNboKGWaQK0sZxxgbS1njCdRpbdZUcBoHgaqhcBo7WvyvrGZzLwABZD0d7UgaVE7JcG4DnX9pqmoAJgsAB5QW6t9PjiNdjr3agcHyD4lywtOPEuEG5IPDGaO+OF7bEfzMOWujq1j2ekN-o486snPXId0KDhihgK6avjo9JcViGLeOYvWNe96mE+rghO+36zeSSuLFB-7fX7ibYltOfnxHuuHBceGqaRlGWfR8YPKN3Ho5tABaBNcgX6vSbjqLfRnOdF3UtdBcOMGpdluB5YHsrVW0NUDqzgNHccmAFBwH9nALGDgPIFmjpzRwIw+A4UvP9Tmmx3qkBqHIXAVUQL3yfqwBOdM-xtTdtmeg6dM5cGzhAXOto4CUW+MNXkQQJpTRmtEOaC1y7fG3pw0ueoD7Q2rm3WcHcSjzUhqjYuQ956KnDlHWOPdn59x-oA+0yhHQgQIKpGAUFtBekPIY8CMBIIaVgvBRCyFgCoTjAZdyeFaY71DIiC8gZW47gCHuA8fEvo1Wyn9bKvj9GhxUS7b0ycPH9XoF7Q8C80oWOXCYsx-i-E-2DtVTS-0XYcNGtGVOCRvYJF9maCkhT4isWCTsC8DSdZQD0egOy+B+C434N4WgHkVxkEgGADIYAAgAApIDVCuJ5Xq-S4DEAgDKFKOUyD+wAJQBFmeIEgYBCQ5lYOQJh0VcrVX9gEckYimjPFoO4cwBDoH8C7uwMQYROC8HGPAHgEApBcB3PARZHAeA8GXO5dIczgDzKgEweQlUbZqHtsQxUVhYZ738DVPi+S6o1ISFbNOcTOEJLKck6JaVYl5V0VA+yDBKCfBgHibKzlXKGVwviaAYVODsHEOQRkxAAjpFYNlfpmVGRkFYMMzoLzTDsqufDZKgrCo4V+XAf5gLgU4VBQAWXBdAKFhCUBLHgEwaWd9EWcntPquATAI7GvNBYTk-iZRQGkBqYA0gHVOv2AePRFLcD2XltGEcsBKBLAQviSgHlun9lldlRZszYFUjOJQegzA2C8qOWQEQiwqBMKmaAcUYEpAbNyoM4ZfwDl4gbFmgFVx5bVklQSFcCz4D7KYa8zKgloDEEwu5V5PB7KzM1RCzKAAJYAHASrRAWHiaK2VpbSAjtERwcBRlUmIJlcQIxVk2GWplLCpA-UAE0FlxQ8s0Mgy4emDCmbM7Q5AFikC4SkBgEBFjsElZytwbLZmkEHYwDgjg7n2SwEB9pGABA5SiJ+jEWIOUd0QuGCsAAxKCDAoIABkYE7B2NHHYuRBCziEm5QcczGRhSvls4gmx3DsH7IyCgo6SrgWIOezSEB11MqchwXZfAr6ug4GkAg4hIi3To2O7wGR4DG2-dqis9kMGrrgGABYvgPl+SvblAsTBGQnpgIyGABY7m4GAyBmFkSejNAAIpZEhdHW6rikWsQ+MAQW6KiQFLSuUolNqSW6OqtCSzdAmA2aZawKO1hnOsISGHGJ9pfMWas4F2zRkADMSLwuedMIHGqtSPHBOLvWRR-rNamZgYqHYUhO6tU9Ol3gmWqtBKnjmPL4Dnw4uK35+LQXXGe1S2iiLXh-ZldFSUSrui2nQMcOJqV2MMjHr+MAMgianPLKudwRCGDjaCBqByOA3Jlw7iIlhIkLKoBhWrdCjpuMOBgp-Y5MIXkKARGiOQAI57O2ZUjXwfl97mhbKGDAbKzby3FrAAB3V5rShwHawFzrRlbX+Oda6vo7qlieqAA) and it solves the problem.

Of course, this is still a bug on our end, but this should get you unstuck in the meantime.

**Comment by andrewbranch** [comment]
Fixed by #43733
